[IO-875] CopyDirectoryVisitor ignores fileFilter

pbaumard opened the pull request:
- Fix [IO-875](https://issues.apache.org/jira/browse/IO-875)
- Fix javadoc
- Added assertions for the copied files and to get a 100% coverage

garydgregory reviewed (changes requested):
@pbaumard 
Please do not remove assertions.

pbaumard commented:
> @pbaumard Please do not remove assertions.

I don't see the point of the removed assertions, which are flagged by [SonarQube IDE](https://www.sonarsource.com/fr/products/sonarlint/) :

"Assertions should not compare an object to itself" ([RSPEC-5863](https://rules.sonarsource.com/java/RSPEC-5863/))

```java
assertEquals(visitFileTree, visitFileTree);
```

"Assertions comparing incompatible types should not be made" ([RSPEC-5845](https://rules.sonarsource.com/java/RSPEC-5845/))

```java
assertNotEquals(visitFileTree, "not");
```

pbaumard commented:
> @pbaumard Please do not remove assertions.

I thought at first that those removed assertions were useless since they were flagged by SonarLint ([RSPEC-5845](https://rules.sonarsource.com/java/RSPEC-5845/) and [RSPEC-5863](https://rules.sonarsource.com/java/RSPEC-5863/)).
I added them back and also added some new assertions to get a 100% coverage.
It could be better to move the equals and hashCode tests to some new method to make the intent clearer. For now I just added some comments.

garydgregory commented:
@pbaumard 

Thank you for your updates. I merged this PR followed up by some cleanups in git master.

TY!
